src/connector/src/source/nexmark/source/generator.rs

```diff
@@ -114,5 +120,10 @@ impl NexmarkEventGenerator {
                 .await;
             }
         }
+
+        // The connector assumes that the stream will never end, so if the `event_num` is hit, we
+        // pend the stream forever.
+        // TODO: should we allow the stream to finish?
+        let () = pending().await;
```

Comment:
I think the stream generator themselves should always end the stream gracefully, as long as they are not infinite streams. Even for those down-streams, they should end themselves as well when the upstream is closed. This way the errors can be propagated to the sink. The system won't be blocking implicitly.

Reply:
We've discussed this issue and decided not to use the `Stream terminate` or `TryStream error` to represent the stream state in RisingWave as we don't want to propagate the error through the network. 🥵 The error yielded in actor internally should be collected by the `Actor` instance and find a way to report it to meta service if possible. cc @fuyufjh 

For the stream reader, I find that the refactor just merged has removed the assumption (as we're all-in async stream), so maybe the workaround can be removed. Note that there's a `select` of source reader and barrier receiver, the source executor will work correctly after the source is gracefully terminated. On error, it will propagate to the actor of this source executor.